Zend/zend_string.c

```diff
@@ -78,57 +88,85 @@ void zend_known_interned_strings_init(zend_string ***strings, uint32_t *count)
 	*count   = known_interned_strings_count;
 }
 
-void zend_interned_strings_init(void)
+static void zend_init_interned_strings_ht(HashTable *interned_strings)
+{
+	zend_hash_init(interned_strings, 1024, NULL, _str_dtor, 1);
+
+	interned_strings->nTableMask = -interned_strings->nTableSize;
+	HT_SET_DATA_ADDR(interned_strings, pemalloc(HT_SIZE(interned_strings), 1));
+	HT_HASH_RESET(interned_strings);
+	interned_strings->u.flags |= HASH_FLAG_INITIALIZED;
+}
+
+#ifdef ZTS
+void zend_interned_strings_init_thread(void)
+{
+	zend_init_interned_strings_ht(&interned_strings_volatile);
+}
+#endif
+
+ZEND_API zend_string *zend_interned_strings_get_empty_string(void)
+{
+	return empty_string;
+}
+
+ZEND_API void zend_interned_strings_init(zend_interned_strings_init_stage storage_type)
 {
-#ifndef ZTS
 	zend_string *str;
 
-	zend_hash_init(&CG(interned_strings), 1024, NULL, _str_dtor, 1);
+	if (ZEND_INTERNED_STRINGS_UNINITIALIZED != zend_strings_storage_type || !(storage_type > ZEND_INTERNED_STRINGS_UNINITIALIZED)) {
+		return;
+	}
 
-	CG(interned_strings).nTableMask = -CG(interned_strings).nTableSize;
-	HT_SET_DATA_ADDR(&CG(interned_strings), pemalloc(HT_SIZE(&CG(interned_strings)), 1));
-	HT_HASH_RESET(&CG(interned_strings));
-	CG(interned_strings).u.flags |= HASH_FLAG_INITIALIZED;
+	zend_init_interned_strings_ht(&interned_strings_permanent);
+	zend_init_interned_strings_ht(&interned_strings_volatile);
+
+	zend_new_interned_string = zend_new_interned_string_permanent_int;
+	zend_interned_strings_snapshot = zend_interned_strings_snapshot_int;
+	zend_interned_strings_restore = zend_interned_strings_restore_int;
 
 	/* interned empty string */
 	str = zend_string_alloc(sizeof("")-1, 1);
 	ZSTR_VAL(str)[0] = '\000';
-	CG(empty_string) = zend_new_interned_string_int(str);
-#endif
+	empty_string = zend_new_interned_string_permanent_int(str);
+#ifndef ZTS
+	CG(empty_string) = empty_string;
```

Comment:
since empty_string is a true global variable,  seems we don't need to store it in CG .